### ISSUE TITLE
Alacritty doesn't work in VirtualBox

### ISSUE THREAD

**Issue by ldanko** [jwilm/alacritty#313]
Alcracitty running on Fedora 25 (on VirtualBox) crashes at startup:
```
$ RUST_BACKTRACE=1 target/release/alacritty 
OpenGL Warning: glXChooseVisual: ignoring attribute 0x22
device_pixel_ratio: 1
width: 1024, height: 768
thread 'main' panicked at 'gl function was not loaded', /home/ldanko/work/alacritty/target/release/build/alacritty-d032f55a418cfed4/out/gl_bindings.rs:20591
stack backtrace:
   1:     0x55d3d2b4752a - std::sys::imp::backtrace::tracing::imp::write::h9c41d2f69e5caabf
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:42
   2:     0x55d3d2b469b1 - std::panicking::default_hook::{{closure}}::h1f61f3c769fffe7a
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:351
   3:     0x55d3d2b45b77 - std::panicking::rust_panic_with_hook::hffbc74969c7b5d87
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:367
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:555
   4:     0x55d3d2a48711 - _ZN3std9panicking11begin_panic17h0a82ee08bc6f08bcE.641
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:517
   5:     0x55d3d2a486cb - alacritty::gl::missing_fn_panic::he1a3e5d8d51883fb
                        at /home/ldanko/work/alacritty/src/lib.rs:85
   6:     0x55d3d2a75df9 - alacritty::display::Display::new::h05668100d7c3e445
                        at /home/ldanko/work/alacritty/src/lib.rs:85
                        at /home/ldanko/work/alacritty/src/renderer/mod.rs:471
                        at /home/ldanko/work/alacritty/src/display.rs:154
   7:     0x55d3d2a3c8bb - alacritty::run::he9434698bec6f695
                        at /home/ldanko/work/alacritty/src/main.rs:75
   8:     0x55d3d2a3be89 - alacritty::main::h3744b8d6b60d0e63
                        at /home/ldanko/work/alacritty/src/main.rs:59
   9:     0x55d3d2a482d0 - main
  10:     0x7faa9a6eb400 - __libc_start_main
  11:     0x55d3d2a0e6e9 - _start
  12:                0x0 - <unknown>

```

**Comment by ldanko** [comment]
It seems to be an error in `gl-rs` crate. One of `gl-rs` examples crashes similarly, see brendanzab/gl-rs#413.

**Comment by mbilker** [comment]
Does something like `glxinfo` work? The backtrace shows that something with OpenGL is not being loaded.

**Comment by ldanko** [comment]
Yes, `glxinfo` works, here is output:
```
glxinfo
name of display: :1
display: :1  screen: 0
direct rendering: Yes
server glx vendor string: Chromium
server glx version string: 1.3 Chromium
server glx extensions:
    GLX_ARB_get_proc_address, GLX_ARB_multisample, 
    GLX_EXT_texture_from_pixmap, GLX_SGIX_fbconfig
client glx vendor string: Chromium
client glx version string: 1.3 Chromium
client glx extensions:
    GLX_ARB_get_proc_address, GLX_ARB_multisample, 
    GLX_EXT_texture_from_pixmap, GLX_SGIX_fbconfig
GLX version: 1.3
GLX extensions:
    GLX_ARB_get_proc_address, GLX_ARB_multisample, 
    GLX_EXT_texture_from_pixmap, GLX_SGIX_fbconfig
OpenGL vendor string: Humper
OpenGL renderer string: Chromium
OpenGL version string: 2.1 Chromium 1.9
OpenGL shading language version string: 4.50
...
...
```
Just look on the commen in linked above issue from `gl-rs`. Author says, that this probable issue in VirtualBox.

**Comment by deathbeam** [comment]
Same issue here, also trying to run Alacritty in VirtualBox, and getting same error.

**Comment by miikkas** [comment]
I'm getting the same error & backtrace. Alacritty worked perfectly up until I installed the VirtualBox guest additions from the VirtualBox menu (I was using the guest addition from the Ubuntu repos before that).

**Comment by OJFord** [comment]
Same issue in VMware.

**Comment by OJFord** [comment]
I previously commented that I had this issue in VMWare too, that was incorrect (user error) - now I'm commenting to say that it Works For Me just fine in VirtualBox and arch.

**Comment by chrisduerr** [comment]
Since this seems to work according to @OJFord and there has been no activity in the last 8 months, I'm closing this issue.

If this problem still exists, please let me know and I can re-open this.

**Comment by miikkas** [comment]
Doesn’t crash anymore, but doesn’t start either. Here are my results with:

- VirtualBox 5.2.12 (with VirtualBox Extension Pack)
- Ubuntu 18.04
    - `uname -a` output: `Linux devvm 4.15.0-20-generic #21-Ubuntu SMP Tue Apr 24 06:16:15 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
- VirtualBox Linux Guest Additions (latest version, via VirtualBox menu) installed



Built with debug mode:
```
$ cargo build
```

Normal startup:
```
$ ./target/debug/alacritty
OpenGL Warning: glXChooseVisual: ignoring attribute 0x22
Alacritty encountered an unrecoverable error:

	Error creating GL context; Couldn't find any pixel format that matches the criterias.

```

Startup using debug mode:
```
$ RUST_BACKTRACE=1 ./target/debug/alacritty 
OpenGL Warning: glXChooseVisual: ignoring attribute 0x22
Alacritty encountered an unrecoverable error:

	Error creating GL context; Couldn't find any pixel format that matches the criterias.

```

`glxinfo` output:
```
$ glxinfo
name of display: :0
OpenGL Warning: vboxCall failed with VBox status code VERR_BUFFER_OVERFLOW

display: :0  screen: 0
direct rendering: Yes
server glx vendor string: Chromium
server glx version string: 1.3 Chromium
server glx extensions:
    GLX_ARB_get_proc_address, GLX_ARB_multisample, 
    GLX_EXT_texture_from_pixmap, GLX_SGIX_fbconfig
client glx vendor string: Chromium
client glx version string: 1.3 Chromium
client glx extensions:
    GLX_ARB_get_proc_address, GLX_ARB_multisample, 
    GLX_EXT_texture_from_pixmap, GLX_SGIX_fbconfig
GLX version: 1.3
GLX extensions:
    GLX_ARB_get_proc_address, GLX_ARB_multisample, 
    GLX_EXT_texture_from_pixmap, GLX_SGIX_fbconfig
OpenGL vendor string: Humper
OpenGL renderer string: Chromium
OpenGL version string: 2.1 Chromium 1.9
OpenGL shading language version string: 4.60 NVIDIA
OpenGL extensions:
    GL_ARB_depth_texture, GL_ARB_draw_buffers, GL_ARB_fragment_program, 
    GL_ARB_fragment_shader, GL_ARB_multisample, GL_ARB_multitexture, 
    GL_ARB_occlusion_query, GL_ARB_pixel_buffer_object, 
    GL_ARB_point_parameters, GL_ARB_point_sprite, GL_ARB_shader_objects, 
    GL_ARB_shader_texture_lod, GL_ARB_shading_language_100, GL_ARB_shadow, 
    GL_ARB_texture_border_clamp, GL_ARB_texture_compression, 
    GL_ARB_texture_cube_map, GL_ARB_texture_env_add, 
    GL_ARB_texture_env_combine, GL_ARB_texture_env_crossbar, 
    GL_ARB_texture_env_dot3, GL_ARB_texture_float, 
    GL_ARB_texture_mirrored_repeat, GL_ARB_texture_non_power_of_two, 
    GL_ARB_texture_rectangle, GL_ARB_transpose_matrix, 
    GL_ARB_vertex_buffer_object, GL_ARB_vertex_program, GL_ARB_vertex_shader, 
    GL_ARB_window_pos, GL_ATI_texture_mirror_once, GL_CR_bounding_box, 
    GL_CR_cursor_position, GL_CR_head_spu_name, GL_CR_performance_info, 
    GL_CR_print_string, GL_CR_readback_barrier_size, GL_CR_saveframe, 
    GL_CR_server_id_sharing, GL_CR_server_matrix, GL_CR_state_parameter, 
    GL_CR_synchronization, GL_CR_tile_info, GL_CR_tilesort_info, 
    GL_CR_window_size, GL_EXT_blend_color, GL_EXT_blend_equation_separate, 
    GL_EXT_blend_func_separate, GL_EXT_blend_minmax, GL_EXT_blend_subtract, 
    GL_EXT_compiled_vertex_array, GL_EXT_draw_range_elements, 
    GL_EXT_fog_coord, GL_EXT_framebuffer_blit, GL_EXT_framebuffer_object, 
    GL_EXT_multi_draw_arrays, GL_EXT_secondary_color, GL_EXT_shadow_funcs, 
    GL_EXT_stencil_two_side, GL_EXT_stencil_two_side, GL_EXT_stencil_wrap, 
    GL_EXT_texture3D, GL_EXT_texture_compression_s3tc, 
    GL_EXT_texture_cube_map, GL_EXT_texture_edge_clamp, 
    GL_EXT_texture_env_add, GL_EXT_texture_env_combine, 
    GL_EXT_texture_env_dot3, GL_EXT_texture_filter_anisotropic, 
    GL_EXT_texture_lod_bias, GL_EXT_texture_object, GL_EXT_texture_sRGB, 
    GL_IBM_rasterpos_clip, GL_IBM_texture_mirrored_repeat, GL_NV_fog_distance, 
    GL_NV_fragment_program, GL_NV_fragment_program2, 
    GL_NV_fragment_program_option, GL_NV_register_combiners, 
    GL_NV_register_combiners2, GL_NV_texgen_reflection, 
    GL_NV_texture_rectangle, GL_NV_vertex_program, GL_NV_vertex_program1_1, 
    GL_NV_vertex_program2, GL_NV_vertex_program2_option, 
    GL_NV_vertex_program3, GL_SGIS_generate_mipmap

1 GLX Visuals
    visual  x   bf lv rg d st  colorbuffer  sr ax dp st accumbuffer  ms  cav
  id dep cl sp  sz l  ci b ro  r  g  b  a F gb bf th cl  r  g  b  a ns b eat
----------------------------------------------------------------------------
0x021 24 tc  0  32  0 r  y y   8  8  8  8 .  .  0 24  8 16 16 16 16  0 0 None

1 GLXFBConfigs:
    visual  x   bf lv rg d st  colorbuffer  sr ax dp st accumbuffer  ms  cav
  id dep cl sp  sz l  ci b ro  r  g  b  a F gb bf th cl  r  g  b  a ns b eat
----------------------------------------------------------------------------
0x021 24 tc  0  32  0 r  y .   8  8  8  8 .  .  0 24  8  0  0  0  0  1 1 None

```

**Comment by chrisduerr** [comment]
That seems like a different issue. Looks like https://github.com/jwilm/alacritty/issues/1268 to me.

**Comment by adam-beck** [comment]
I have to apologize as I won't really be contributing to helping solve this problem as I don't understand half of the comments. However, I just tried this with 

- VirtualBox 5.2.0
- Ubuntu 18.04.1 LTS
- virtualbox-guest-x11 / VBox_GAs_5.2.20

And with the additions installed (tried each separately) alacritty would not open with the same message as @miikkas 

> Error creating GL context; Couldn't find any pixel format that matches the criteria.

I removed the additions and both times alacritty would start properly.

**Comment by PetrGlad** [comment]
I am facing the same issue (Xubuntu in VirtualBox, with guest additions, on Windows 10 host), I compiled from latest sources hence patch for #1268 is already there. So that patch does not help, I am still getting "Couldn't find any pixel format that matches the criteria."

However I tried to modify  get_platform_window(...) with `.with_transparency(false)` so the message is gone, but now it stops with 
```
$ RUST_BACKTRACE=1 cargo run 
    Finished dev [unoptimized + debuginfo] target(s) in 0.32s                                                                                                                                                                                                                  
     Running `target/debug/alacritty`
OpenGL Warning: glXChooseVisual: ignoring attribute 0x22
thread 'main' panicked at 'gl function was not loaded', /home/petr/work/etc/alacritty/target/debug/build/alacritty-84ae4f16633b9092/out/gl_bindings.rs:20624:13
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:477
   5: std::panicking::begin_panic
             at libstd/panicking.rs:411
   6: alacritty::gl::missing_fn_panic
             at ./target/debug/build/alacritty-84ae4f16633b9092/out/gl_bindings.rs:20624
   7: alacritty::gl::GenVertexArrays
             at ./target/debug/build/alacritty-84ae4f16633b9092/out/gl_bindings.rs:1946
   8: alacritty::renderer::QuadRenderer::new
             at src/renderer/mod.rs:486
   9: alacritty::display::Display::new
             at src/display.rs:153
  10: alacritty::run
             at src/main.rs:129
  11: alacritty::main
             at src/main.rs:85
  12: std::rt::lang_start::{{closure}}
             at libstd/rt.rs:74
  13: std::panicking::try::do_call
             at libstd/rt.rs:59
             at libstd/panicking.rs:310
  14: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:103
  15: std::rt::lang_start_internal
             at libstd/panicking.rs:289
             at libstd/panic.rs:392
             at libstd/rt.rs:58
  16: std::rt::lang_start
             at libstd/rt.rs:74
  17: main
  18: __libc_start_main
  19: _start
```
The new program window is created briefly but then alacritty stops.

**Comment by chrisduerr** [comment]
@PetrGlad The patch hasn't landed in master yet, since it's not just an improvement across the board and more of a workaround.

The other error you've posted sounds like there's problems with OpenGL support. Alacritty requires OpenGL to render its windows, so make sure that's properly supported and drivers are installed.

**Comment by mbilker** [comment]
I have been working with OpenGL in another Rust program and I agree with @chrisduerr. Make sure your graphics drivers are also *up to date*, especially with older hardware. I have an older AMD Radeon driver installed on my computer, but upgrading the driver allowed gfx-rs to load up.

Note: I was not using Alacritty, but it was a relatively simple OpenGL program that still triggered the `gl function was not loaded` error.

**Comment by PetrGlad** [comment]
@chrisduerr My impression was that following in Window::new
```
let window = create_gl_window(window_builder.clone(), &event_loop, false)
            .or_else(|_| create_gl_window(window_builder, &event_loop, true))?;
```
was meant to be a workaround for the problem in #1268 as it tries to create window with and without `srgb` format.

I believe "drivers" in my case are VirtualBox's guest additions. They are installed, match VirtualBox version, and "3d acceleration" is enabled in virtual machine's settings. I think not much else I can do further unless there are some additional libraries that can be installed to fix this.
Although I have almost no GL programming experience, my guess is that missing function is not implemented.

Out of curiosity I tried to run `glutin` examples (same environment, on master, 1ad6560860d), and 
* `fullscreen`, `multiwindow`, `window` - run successfully
* `sharing` - fails with `thread 'main' panicked at 'not yet implemented', src/platform/linux/x11.rs:135:72`
* `transparent` - fails with `thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: NoAvailablePixelFormat', libcore/result.rs:1009:5`

**Comment by PetrGlad** [comment]
glxinfo for virtualbox, just in case.
[virtualbox-glxinfo.txt](https://github.com/jwilm/alacritty/files/2652770/virtualbox-glxinfo.txt)

**Comment by mbilker** [comment]
Thank you for including the stack trace. It appears you are missing `glGenVertexArrays` (it is used [here](https://github.com/jwilm/alacritty/blob/f57bd6e12f7e94b767b34a208c66dc88c93f704e/src/renderer/mod.rs#L486)). That function is only available on OpenGL version 3.0 and newer. Your `glxinfo` report indicates only OpenGL 2.1 support.

**Comment by mbilker** [comment]
There are probably other GL functions used that are only available on OpenGL 3.0 and newer.

According to [](https://askubuntu.com/questions/858407/how-to-update-to-latest-opengl-version-on-virtualbox-ubuntu-linux-machine), VirtualBox only supports up to OpenGL 2.1 so Alacritty will need to use a different OpenGL function if VirtualBox compatibility is desired with acceleration.

**Comment by chrisduerr** [comment]
Thanks for looking into this @mbilker. Please follow #128 for tracking of minimal OpenGL version. Most people aren't affected by this, so it's not a priority for me right now. All contributions torwards this would be much appreciated though!
